src/roster/views.py

```diff
@@ -1495,6 +1497,23 @@ def get(self, request, *args, **kwargs):
         # TODO: (JS) fill these in as necessary.
         self.panels = dict()
         self.panel_data = dict()
+
+        # Student Activity panel
+        self.panels['activity'] = self.clusive_student
+        if self.panels['activity']:
+            students_reading_data = Paradata.reading_data_for_period(self.clusive_user.current_period, days=self.days, sort='name')
+            target_reading_data = {}
+            for one_reading_data in students_reading_data:
```

Comment:
I'm wondering if it's worth it to make this loop a class function of Paradata, something like:
```
@classmethod
def reading_data_for_student_period(student: ClusiveUser, period: Period, days, sort)
```

Reply:
I agree it would be helpful to have such a class function.

I'm thinking the code might be more efficient by adding an optional argument `username` to reading_data_for_period(). If `username == None`, look up for all students, otherwise, only look up for the given student.

The current function name may need to be adjusted to reflect this new feature.

Let me know what you think. Thanks.

Reply:
Either way would be a big improvement over extracting the data for all students and then throwing away everything that's not the current student.

We don't want _everything_ that is in `reading_data_for_student_period` though. In particular, that method has code to collapse readings that are only a small amount of time into a single "other" record. We don't want to do that for the details page. We don't need to calculate the time percentages at all.

The most elegant would be split it into two methods - a new class method that pulls basic data either for a full class or a single student - and then change `reading_data_for_student_period` to call that and do its extra processing, and the student details view method to call that and _not_ do the extra processing.

Reply:
Either way would be a big improvement over extracting the data for all students and then throwing away everything that's not the current student.

We don't want _everything_ that is in `reading_data_for_student_period` though. In particular, that method has code to collapse readings that are only a small amount of time into a single "other" record. We don't want to do that for the details page. We don't need to calculate the time percentages at all.

The most elegant would be split it into two methods - a new class method that pulls basic data either for a full class or a single student - and then change `reading_data_for_student_period` to call that and do its extra processing, and the student details view method to call that and _not_ do the extra processing.